### PR TITLE
Improve DMD's JSON output: Always output visibility & give a special kind to module ctors

### PR DESCRIPTION
**Pull request by Geod24**
See each commit for details.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
Changelog?

**Comment by Geod24** [comment]
It's the first two files in the diff ?

**Comment by Geod24** [comment]
I think think needs some fixes from `dpl-docs` though. CC @adamdruppe @CyberShadow

**Comment by adamdruppe** [comment]
dpl-docs is actually not me, I only do the website domain dpldocs.info

hopefully someday we'll use my docgen here too but right now there's no connection here; I don't use dmd's json output and dmd doesn't use my generator.

(this comes up a lot, the names are so similar i sometimes make the mistake too, but yeah, it is separate.)

File: src/dmd/json.d [hunk 1]
```diff
@@ -416,13 +416,23 @@ public:
     {
         if (s.isModule())
             return;
-        if (!s.isTemplateDeclaration()) // TemplateDeclaration::kind() acts weird sometimes
+        if (s.isStaticCtorDeclaration())
+        {
+            property("kind", s.isSharedStaticCtorDeclaration()
+                     ? "sharedStaticConstructor" : "staticConstructor");
```
Comment (CyberShadow) [inline src/dmd/json.d:422]: The current convention for multiple-word kinds seems to be separating the words by spaces, and not using camel case.

**Comment by CyberShadow** [comment]
I don't know why dpldocs/ddox is failing. The second problem I see in the logs is that it's built without debug information, so it's difficult to tell where it's failing exactly.

By the way, this change could have been done in a completely non-breaking way, by adding new fields (e.g. boolean flags) to objects instead of adding new property kinds, which will require updates in all downstream consumers. I don't know if it's actually worthwhile to do that, but it would at least avoid the problem at hand.

File: src/dmd/json.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/json.d]: Indeed, since we feed "kind" verbatim. Updated.

**Comment by Geod24** [comment]
> By the way, this change could have been done in a completely non-breaking way, by adding new fields (e.g. boolean flags) to objects instead of adding new property kinds, which will require updates in all downstream consumers. 

Indeed, but things will break the same when the names are changed, and the API would be uglier.

**Comment by CyberShadow** [comment]
Why would we want to change the internal name?

Things like `emplace` already rely on the internal names, don't they?

**Comment by Geod24** [comment]
Well, to be more precise, I wanted to make them less prominent.
The original work was done in [a PR trying to remove magic identifiers from the scope](https://github.com/dlang/dmd/pull/10834). In order to do so, I had to go through the list of things that used those identifiers, and the JSON generator was one of them.

**Comment by Geod24** [comment]
Fixed by removing the breaking change, CC @thewilsonator

**Pull request merged**